prompt.go

```diff
@@ -346,18 +343,53 @@ func (p *Prompt) tearDown() {
 	p.renderer.TearDown()
 }
 
-func (p *Prompt) turnOffReverseSearch(eraseLinesCnt int) {
-	for i := 0; i < eraseLinesCnt; i++ {
-		p.renderer.out.CursorUp(1)
-		p.renderer.out.EraseLine()
+// enableReverseSearch enables reverse-search mode.
+func (p *Prompt) enableReverseSearch() {
+	if !p.reverseSearchOpt {
+		return
 	}
 
-	p.renderer.out.CursorBackward(maxBackwardColumn)
+	if !p.renderer.reverseSearch.enabled {
+		p.buf = NewBuffer()
+	}
+	p.renderer.reverseSearch.onActivate(p.history)
+}
 
+// disableReverseSearch disables reverse-search mode,
+// sets history pointer to the last matched entry.
+func (p *Prompt) disableReverseSearch() {
+	if !p.reverseSearchOpt {
+		return
+	}
+	matchedIndex := p.renderer.reverseSearch.matchedIndex
+	matchedCmd := p.renderer.reverseSearch.matchedCmd
+	p.history.Clear()
 	p.buf = NewBuffer()
-	p.buf.setText(p.history.lastReverseFinded)
 
-	p.renderer.reverseSearchEnabled = false
-	p.renderer.searchPrefixIsSet = false
-	p.history.lastReverseFinded = ""
+	if matchedIndex != -1 {
+		p.history.selected = matchedIndex
+		p.history.SetCurrentCmd(matchedCmd)
+		p.buf.InsertText(matchedCmd, false, true)
+	}
+
+	p.renderer.reverseSearch.onDeactivate()
+}
+
+// saveCursor saves current cursor state.
+func (p *Prompt) saveCursor() {
+	p.prevCursor = p.curCursor
+}
+
+// PushToHistory pushes entry to history, if such option is enabled.
+func (p *Prompt) PushToHistory(input string) error {
+	if p.pushHistory {
```

Comment:
Probably we need a test for the code.
```suggestion
	if !p.pushHistory {
```